### ISSUE TITLE
S3453: FP with SafeHandle

### ISSUE THREAD

**Issue by ThorstenReichert** [SonarSource/sonar-dotnet#3329]
### Description

When creating a custom SafeHandle, it is possible to only have a private constructor. This SafeHandle still can be instantiated by the PInvoke Marshaler, as shown in the snippet below.

### Repro steps

```csharp
public class MySafeHandle : SafeHandleZeroOrMinusOneIsInvalid
{
    private MySafeHandle() : base(true)
    {
    }

    protected override bool ReleaseHandle()
    {
        throw new NotImplementedException();
    }
}

public static class NativeMethods
{
    [DllImport("Some.dll")]
    private static extern MySafeHandle CreateSafeHandle();
}
```

### Expected behavior

No warning.

### Actual behavior

Warning on `MySafeHandle`: S3453: This class can't be instantiated; make its constructor 'public'.

### Related information

* C# Plugins Version 8.6.1.17183
* Visual Studio Version 16.5.4
* dotnet version .NET Core 3.1
* Operating System Windows 10

**Comment by pavel-mikula-sonarsource** [comment]
Hi @ThorstenReichert,

We confirm that this is a False Positive in this special case, but unfortunately we're not going to fix it. 
* It is a good practice to have the constructor private here, but it's not required by design and it can be public. 
* Detecting this issue would mean a significant performance penalty for the rule.
* `DllImport` static function can be placed in external assembly and fix would not handle that scenario.

**Comment by ThorstenReichert** [comment]
Hi @pavel-mikula-sonarsource,

Thank you for the clarification. I understand that scanning for usages of a SafeHandle in DllImports would not be feasable.

On the other hand, would it be possible to disable this rule for SafeHandles (I'd guess checking if a class derives from SafeHandle is relatively cheap)? Or do you think this, in turn, would lead to too many false negatives?

**Comment by pavel-mikula-sonarsource** [comment]
Hi @ThorstenReichert,

That's a very good argument that I overlooked. So I'm reopening the issue.
